tools/hlslpm/issue.py

```diff
@@ -0,0 +1,347 @@
+    
+    def getContentsFor(self, issue) -> List[str]:
+        (_, data, _) = split_body(self.body)
+        data = parse_data(data)
+
+        for s in data.sections:
+            reference = self.convertReferenceToResourcePath(s.getReferenceFromTitle())
+            if reference == issue.issue_resourcePath:
+                return s.contents
+            
+        return []
+
+
+
+
+def split_body(body: str):
+    """An issue's body is made up of three sections:
+
+    * pre
+    * data
+    * post
+
+    Data is identified by a line starting with "## Milestones" or "##
+    Workstreams".
+
+    The post section starts with a horizontal rule (at least 3 '-' characters).
+    """
+    pre = []
+    data = []
+    post = []
+
+    state = "pre"
+
+    for line in body.splitlines():
+        if state == "pre":
+            if line.startswith("## Milestones") or line.startswith("## Workstreams"):
+                state = "data"
+            else:
+                pre.append(line)
+        if state == "data":
+            if line.startswith('---'):
+                state = "post"
+            else:
+                data.append(line)
+        if state == "post":
+            post.append(line)
+
+    return (pre, data, post)
+
+
+def rebuild_body(pre: List[str], body: List[str], post: List[str]):
+    return "\n".join(pre + body + post)
+
+
+def parse_data(data: List[str]) -> IssueData:
+    """
+    The data section can list milestones or workstreams, depending on the title
+    that starts with "##".  Anything before this title is ignored.
+
+    Then each milestone/workstream is identified with "###", and the contents of
+    the section is preserved.
+    """
+    d = IssueData()
+
+    if not data:
+        return d
+
+    currentSection = None
+
+    for line in data:
+        if d.type == None:
+            m = re.match("## (Milestones|Workstreams)", line)
+            if m:
+                d.type = m[1]
+            continue
+
+        # Check for a new section
+        m = re.match("### (.*)", line)
+        if m:
+            if currentSection:
+                d.sections.append(currentSection)
+            currentSection = IssueSection()
+            currentSection.title = m[1].strip()
+        else:
+            if not currentSection:
+                # create a dummy section for stray lines before a section header
+                currentSection = IssueSection()
+            currentSection.contents.append(line)
+
+    if currentSection:
+        d.sections.append(currentSection)
+
+    return d
+
+
+def rebuild_data(data: IssueData) -> List[str]:
+    body = [f"## {data.type}"]
+
+    for section in data.sections:
+        if section.title:
+            body.append(f"### {section.title}")
+
+        for line in section.contents:
+            body.append(line)
+            
+        if body[-1] != "":
+            body.append("")
+
+    return body
+
+class Issues:
+    all_issues: Dict[str, Issue]
```

Comment:
I'm having trouble understanding why a dictionary is used to hold the issues instead of a list. It looks like the key is `i.issue_resourcePath`, but that seems like it could just be part of the `Issue`? I'm sure there is a good reason, but I think adding some explanation about the data structure here would help.

Reply:
In the case of all_issues it is because we need to be able to look them up by resourcePath, which I'd argue is self documenting it by being a Dict.

milestones and workstreams I don't think need to be a dict.